internal/charmstore/store.go

```diff
@@ -719,40 +723,92 @@ var seriesScore = map[string]int{
 	"utopic":  4,
 	"vivid":   5,
 	"wily":    6,
+	// When we find a multi-series charm (no series) we
+	// will always choose it in preference to a series-specific
+	// charm
+	"": 5000,
 }
 
 // EntitiesQuery creates a mgo.Query object that can be used to find
 // entities matching the given URL. If the given URL has no user then
 // the produced query will only match promulgated entities.
 func (s *Store) EntitiesQuery(url *charm.Reference) *mgo.Query {
-	if url.User != "" && url.Series != "" && url.Revision != -1 {
-		// Find a specific owned entity, for instance ~who/utopic/django-42.
-		return s.DB.Entities().FindId(url)
-	}
-	if url.Series != "" && url.Revision != -1 {
-		// Find a specific promulgated entity, for instance utopic/django-42.
+	entities := s.DB.Entities()
+	// TODO this could be phrased more efficiently.
```

Comment:
This seems unnecessarily hard to follow. Which I expect is on the basis of some kind of theoretical optimizations. How about: 
```go
query := bson.D{{
	"name", url.Name,
}}
if url.User == "" {
	if url.Revision > -1 {
		query = append(query, bson.DocElem{"promulgated-revision", url.Revision})
	} else {
		query = append(query, bson.DocElem{"promulgated-revision", bson.D{{"$gt", -1}}})
	}
} else {
	query = append(query, bson.DocElem{"user", url.User})
	if url.Revision > -1 {
		query = append(query, bson.DocElem{"revision", url.Revision})
	}
}
if url.Series != nil {
	query = append(query, bson.DocElem{"supportedseries", url.Series})
}
return entities.Find(query)
```